test-framework/arquillian/src/main/java/io/quarkus/arquillian/QuarkusDeployableContainer.java

```diff
@@ -0,0 +1,268 @@
+            }
+
+            Path testClassesLocation;
+            try {
+                testClassesLocation = PathTestHelper.getTestClassesLocation(testJavaClass);
+            } catch (Exception e) {
+                // TCK tests are usually located in a dependency jar
+                testClassesLocation = new File("target/test-classes").toPath();
+            }
+
+            URLClassLoader appCl;
+            try {
+                BootstrapClassLoaderFactory clFactory = BootstrapClassLoaderFactory.newInstance()
+                        .setAppClasses(appLocation)
+                        .setParent(testJavaClass.getClassLoader())
+                        .setOffline(PropertyUtils.getBooleanOrNull(BootstrapClassLoaderFactory.PROP_OFFLINE))
+                        .setLocalProjectsDiscovery(
+                                PropertyUtils.getBoolean(BootstrapClassLoaderFactory.PROP_WS_DISCOVERY, true));
+                for (Path library : libraries) {
+                    clFactory.addToClassPath(library);
+                }
+                appCl = clFactory.newDeploymentClassLoader();
+
+            } catch (BootstrapException e) {
+                throw new IllegalStateException("Failed to create the boostrap class loader", e);
+            }
+
+            RuntimeRunner runner = RuntimeRunner.builder()
+                    .setLaunchMode(LaunchMode.TEST)
+                    .setClassLoader(appCl)
+                    .setTarget(appLocation)
+                    .setFrameworkClassesPath(testClassesLocation)
+                    .addChainCustomizers(customizers)
+                    .addAdditionalArchives(libraries)
+                    .build();
+
+            runner.run();
+            runtimeRunner.set(runner);
+
+            // Instantiate the real test instance
+            Object actualTestInstance = Class
+                    .forName(testJavaClass.getName(), true, Thread.currentThread().getContextClassLoader()).newInstance();
+            if (actualTestInstance != null) {
+                TestInjectionManager.inject(actualTestInstance);
+            }
+            testInstance = actualTestInstance;
+
+        } catch (Exception e) {
+            throw new DeploymentException("Unable to start the runtime runner", e);
+        }
+
+        ProtocolMetaData metadata = new ProtocolMetaData();
+        URI uri = URI.create(TestHTTPResourceManager.getUri());
+        metadata.addContext(new HTTPContext(uri.getHost(), uri.getPort()));
+        return metadata;
+    }
+
+    @Override
+    public void undeploy(Archive<?> archive) throws DeploymentException {
+        testInstance = null;
+        RuntimeRunner runner = runtimeRunner.get();
+        if (runner != null) {
+            try {
+                runner.close();
+            } catch (IOException e) {
+                throw new DeploymentException("Unable to close the runtime runner", e);
+            } finally {
+                if (deploymentLocation.get() != null) {
```

Comment:
Wondering if we should have an option to keep the deployed directory?

I know I appreciated having one when working with Arquillian in the past.

Reply:
You mean 

```xml
<engine>
  <property name="deploymentExportPath">target/deployments</property>
</engine>
```

I belive that should work out of the box.

Reply:
`deploymentExportPath` works but note that it only exports the test archive. E.g. in case of quarkus the test archive does not contain the wiring classes and other stuff. Currently, if a deployment fails the tmp directory is not removed from the `target` directory. But it's more like a bug ;-)